addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java

```diff
@@ -0,0 +1,182 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.RawType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link GenericThingState} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Martin S. Eskildsen - Initial contribution
+ */
+public abstract class GenericThingState implements ChannelStateChangePublisher {
+
+    private Logger logger = LoggerFactory.getLogger(GenericThingState.class);
+
+    final AtomicBoolean allowRefresh = new AtomicBoolean(true);
+
+    private ChannelStateChangeSubscriber subscriber = null;
+
+    // Keeps track of current channel refresh status of subscribed channels
+    private Map<String, GenericChannelState> subscriptions = new ConcurrentHashMap<String, GenericChannelState>();
+
+    public GenericThingState(ChannelStateChangeSubscriber subscriber) {
+        this.subscriber = subscriber;
+    }
+
+    protected abstract void recalculate();
+
+    public void onChannelChanged(ChannelUID channelUID) {
+    }
+
+    @Override
+    public void addChannel(ChannelUID channelUID) {
+        try {
+            if (!subscriptions.containsKey(channelUID.getId())) {
+                subscriptions.put(channelUID.getId(), createChannelSubscription(channelUID));
+            }
+        } catch (
+
+        Exception ex) {
+            logger.error("{}: context='subscribe' - Exception occurred when subscribing to channel '{}'", "<UNKNOWN>",
+                    channelUID.getId());
+        }
+    }
+
+    protected GenericChannelState createSubscriptionStringType(ChannelUID channelUID) {
+        return new ChannelStringType(channelUID, this, subscriber);
+
+    }
+
+    protected GenericChannelState createSubscriptionOnOffType(ChannelUID channelUID) {
+        return new ChannelOnOffType(channelUID, this, subscriber);
+    }
+
+    protected GenericChannelState getChannelStateHandler(String channelId) {
+        return subscriptions.get(channelId);
+
+    }
+
+    public GenericChannelState getChannelStateHandler(ChannelUID channelUID) {
+        return getChannelStateHandler(channelUID.getId());
+
+    }
+
+    protected GenericChannelState createSubscriptionRawType(ChannelUID channelUID) {
+        return new ChannelRawType(channelUID, this, subscriber);
+    }
+
+    @Override
+    public void subscribe(ChannelUID channelUID/* , ChannelStateChangeSubscriber subscriber */) {
+        try {
+            if (getChannelStateHandler(channelUID) == null) {
+                addChannel(channelUID/* , subscriber */);
+            }
+            getChannelStateHandler(channelUID).subscribe();
+
+        } catch (
+
+        Exception ex) {
```

Comment:
Fix indentation and please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.